### ISSUE TITLE
undefined reference to `isal_adler32'

### ISSUE THREAD

**Issue by bolunfeng** [intel/isa-l#82]
On the windows platform, I manually compile .o one by one,for example:
gcc  -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -D__EXTENSIONS__=1 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D_POSIX_PTHREAD_SEMANTICS=1 -D_TANDEM_SOURCE=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_AS_KNOWS_AVX512=1 -DHAVE_LIMITS_H=1 -DHAVE_STDINT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_STDLIB_H=1 -DHAVE_MALLOC=1 -DHAVE_MEMMOVE=1 -DHAVE_MEMSET=1 -DHAVE_GETOPT=1 -I.    -Wall -Wchar-subscripts -Wformat-security -Wnested-externs -Wpointer-arith -Wshadow -Wstrict-prototypes -Wtype-limits  -I ./include/ -I ./erasure_code -I ./crc -I ./igzip -I ./tests/fuzz -I ./examples/ec  -g -O2 -MT erasure_code/ec_base.lo -MD -MP -MF $depbase.Tpo -c -o erasure_code/ec_base.lo erasure_code/ec_base.c

nasm -f elf64 -I ./include/ -I ./erasure_code -I ./crc -I ./igzip -I ./tests/fuzz -I ./examples/ec  -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -D__EXTENSIONS__=1 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D_POSIX_PTHREAD_SEMANTICS=1 -D_TANDEM_SOURCE=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_AS_KNOWS_AVX512=1 -DHAVE_LIMITS_H=1 -DHAVE_STDINT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_STDLIB_H=1 -DHAVE_MALLOC=1 -DHAVE_MEMMOVE=1 -DHAVE_MEMSET=1 -DHAVE_GETOPT=1 -g -O2  erasure_code/gf_vect_mul_avx.asm -o erasure_code/gf_vect_mul_avx.o

.........

 and then generate libisal.a with the ar tool. But when my executor links this library, I get the following error:

../../lib/win/gcc/debug/intel/isa-l/lib/libisal.a(igzip.o): In function `isal_adler32_bam1':
D:\software\gt\3thparty\intel\isa-l/igzip/igzip.c:151: undefined reference to `isal_adler32'
../../lib/win/gcc/debug/intel/isa-l/lib/libisal.a(igzip.o): In function `update_checksum':
D:\software\gt\3thparty\intel\isa-l/igzip/igzip.c:164: undefined reference to `crc32_gzip_refl'

However, I checked the symbol table for libisal.a, which should contain this function:
nm libisal.a

igzip.o:
0000000000000300 t flush_write_buffer
                 U gzip_hdr
                 U gzip_hdr_bytes
                 U gzip_trl_bytes
                 U hufftables_default
                 U hufftables_static
                 U isal_adler32
00000000000009e0 T isal_adler32_bam1

igzip_multibinary.o:
000000000000067a T isal_adler32
0000000000000680 t isal_adler32_dispatch_init
0000000000000060 d isal_adler32_dispatched
0000000000000675 t isal_adler32_mbinit

 In other words, the igzip_multibinary.o file contains the isal_adler32  function. But why can't find this link in the end?

**Comment by gbtucker** [comment]
If this is for windows you should have the yasm flags -f win64. I don't
know if this is the issue why your symbols are not found.

**Comment by bolunfeng** [comment]
@gbtucker  yes ,maybe you are right !  
my command is:
nasm -f elf64 -I ./include/ -I ./erasure_code -I ./crc -I ./igzip -I ./tests/fuzz -I ./examples/ec -DPACKAGE_NAME=\"libisal\" -DPACKAGE_TARNAME=\"isa-l\" -DPACKAGE_VERSION=\"2.26.0\" "-DPACKAGE_STRING=\"libisal 2.26.0\"" -DPACKAGE_BUGREPORT=\"sg.support.isal@intel.com\" -DPACKAGE_URL=\"http://01.org/storage-acceleration-library\" -DPACKAGE=\"isa-l\" -DVERSION=\"2.26.0\" -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -D__EXTENSIONS__=1 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D_POSIX_PTHREAD_SEMANTICS=1 -D_TANDEM_SOURCE=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_AS_KNOWS_AVX512=1 -DHAVE_LIMITS_H=1 -DHAVE_STDINT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_STDLIB_H=1 -DHAVE_MALLOC=1 -DHAVE_MEMMOVE=1 -DHAVE_MEMSET=1 -DHAVE_GETOPT=1 -g -O2  igzip/adler32_avx2_4.asm -o igzip/adler32_avx2_4.o 


I use nasm -f elf64 ............

**Comment by bolunfeng** [comment]
@gbtucker  I get the error:
d:intel\isa-l>yasm -f win64 -I ./include/ -I ./erasure_code -I ./crc -I ./igzip -I ./tests/fuzz -I ./examples/ec  -DSTDC_HEADERS=1 -DHAVE_SYS_TYPES_H=1 -DHAVE_SYS_STAT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_MEMORY_H=1 -DHAVE_STRINGS_H=1 -DHAVE_INTTYPES_H=1 -DHAVE_STDINT_H=1 -DHAVE_UNISTD_H=1 -D__EXTENSIONS__=1 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D_POSIX_PTHREAD_SEMANTICS=1 -D_TANDEM_SOURCE=1 -DHAVE_DLFCN_H=1 -DLT_OBJDIR=\".libs/\" -DHAVE_AS_KNOWS_AVX512=1 -DHAVE_LIMITS_H=1 -DHAVE_STDINT_H=1 -DHAVE_STDLIB_H=1 -DHAVE_STRING_H=1 -DHAVE_STDLIB_H=1 -DHAVE_MALLOC=1 -DHAVE_MEMMOVE=1 -DHAVE_MEMSET=1 -DHAVE_GETOPT=1 -O2  erasure_code/gf_vect_dot_prod_avx512.asm -o erasure_code/gf_vect_dot_prod_avx512.o
erasure_code/gf_vect_dot_prod_avx512.asm:136: error: undefined symbol `zmm5' (first use)
erasure_code/gf_vect_dot_prod_avx512.asm:136: error:  (Each undefined symbol is reported only once.)
erasure_code/gf_vect_dot_prod_avx512.asm:139: error: undefined symbol `gf_vect_dot_prod_avx512.len_lt_64' (first use)
erasure_code/gf_vect_dot_prod_avx512.asm:142: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:148: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:151: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:152: error: undefined symbol `zmm0' (first use)
erasure_code/gf_vect_dot_prod_avx512.asm:153: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:155: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:158: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:159: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:161: error: undefined symbol `zmm3' (first use)
erasure_code/gf_vect_dot_prod_avx512.asm:162: error: undefined symbol `zmm4' (first use)
erasure_code/gf_vect_dot_prod_avx512.asm:162: error: undefined symbol `zmm1' (first use)
erasure_code/gf_vect_dot_prod_avx512.asm:163: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:164: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:169: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:173: error: undefined symbol `vmovdqu8.loop64' (first use)
erasure_code/gf_vect_dot_prod_avx512.asm:189: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:195: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:198: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:199: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:200: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:202: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:204: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:205: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:206: error: instruction expected after label
erasure_code/gf_vect_dot_prod_avx512.asm:210: error: instruction expected after label

**Comment by gbtucker** [comment]
yasm doesn't support avx512 yet.  You should remove -DHAVE_AS_KNOWS_AVX512

**Comment by bolunfeng** [comment]
@gbtucker  thank. The above question should have been resolved. But still in the end when my program links this library, it is still a problem:

../../lib/win/gcc/debug/intel/isa-l/lib/libisal.a(igzip_multibinary.o):igzip/igzip_multibinary.asm:(.text+0x51d): undefined reference to `encode_deflate_icf_06'
../../lib/win/gcc/debug/intel/isa-l/lib/libisal.a(igzip_multibinary.o):igzip/igzip_multibinary.asm:(.text+0x5be): undefined reference to `set_long_icf_fg_06'
../../lib/win/gcc/debug/intel/isa-l/lib/libisal.a(igzip_multibinary.o):igzip/igzip_multibinary.asm:(.text+0x65f): undefined reference to `gen_icf_map_lh1_06'
../../lib/win/gcc/debug/tbb_debug/libtbb_debug.a(task.o): duplicate section `.rdata$_ZTIN3tbb4taskE[_ZTIN3tbb4taskE]' has different size
../../lib/win/gcc/debug/tbb_debug/libtbb_debug.a(arena.o): duplicate section `.rdata$_ZTIN3tbb4taskE[_ZTIN3tbb4taskE]' has different size
../../lib/win/gcc/debug/tbb_debug/libtbb_debug.a(scheduler.o): duplicate section `.rdata$_ZTIN3tbb4taskE[_ZTIN3tbb4taskE]' has different size
../../lib/win/gcc/debug/tbb_debug/libtbbmalloc_debug.a(tbbmalloc.o): In function `DllMain':
D:/software/gt_compress/3thparty/tbb44_20160413oss/src/tbbmalloc/tbbmalloc.cpp:88: multiple definition of `DllMain'
../../lib/win/gcc/debug/tbb_debug/libtbb_debug.a(tbb_main.o):D:/software/gt_compress/3thparty/tbb44_20160413oss/src/tbb/tbb_main.cpp:247: first defined here
../../lib/win/gcc/debug/intel/isa-l/lib/libisal.a(igzip_multibinary.o):igzip/igzip_multibinary.asm:(.text+0x51d): undefined reference to `encode_deflate_icf_06'
../../lib/win/gcc/debug/intel/isa-l/lib/libisal.a(igzip_multibinary.o):igzip/igzip_multibinary.asm:(.text+0x5be): undefined reference to `set_long_icf_fg_06'
../../lib/win/gcc/debug/intel/isa-l/lib/libisal.a(igzip_multibinary.o):igzip/igzip_multibinary.asm:(.text+0x65f): undefined reference to `gen_icf_map_lh1_06'
collect2.exe: error: ld returned 1 exit status

**Comment by bolunfeng** [comment]
@gbtucker  thank you very much. I have solved this problem. The reason is that I used -DHAVE_AS_KNOWS_AVX512 when compiling igzip_multibinary.asm. You really helped me a lot.
